File: database-driver-api/src/main/java/de/lino/database/json/file/FileProvider.java
package de.lino.database.json.file;

import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.nio.file.Path;
import java.util.Collection;

/**
 * Abstraction over the filesystem operations used throughout the driver (creating, deleting,
 * renaming and copying files and directories), so that {@link de.lino.database.json.JsonDocument}
 * and the various providers do not depend on a concrete filesystem implementation directly.
 * <p>
 * A single, globally accessible instance is installed via {@link #setInstance(FileProvider)} and
 * retrieved through the generated {@code getInstance()} accessor.
 */
@Getter
public abstract class FileProvider {

    /**
     * The globally accessible instance of this database, installed via
     * {@link #setInstance(FileProvider)} and exposed through the generated
     * {@code getInstance()} accessor.
     */
    @Getter
    @Nullable
    protected static FileProvider instance;

    /**
     * Installs the given database as the globally accessible instance returned by the generated
     * {@code getInstance()} accessor.
     *
     * @param instance the database instance to install
     */
    protected static void setInstance(@NotNull FileProvider instance) {
        FileProvider.instance = instance;
    }

    /**
     * Deletes the given file, if it exists.
     *
     * @param file the file to delete
     */
    public abstract void deleteFile(@NotNull File file);

    /**
     * Deletes the file at the given path, if it exists.
     *
     * @param file the path of the file to delete
     */
    public abstract void deleteFile(@NotNull Path file);

    /**
     * Creates the given file, including any missing parent directories, if it does not exist yet.
     *
     * @param file the path of the file to create
     */
    public abstract void createFile(@NotNull Path file);

    /**
     * Creates the given file, including any missing parent directories, if it does not exist yet.
     *
     * @param file the file to create
     */
    public abstract void createFile(@NotNull File file);

    /**
     * Recreates the given file, first deleting it if present and then creating it anew.
     *
     * @param file the path of the file to update
     */
    public abstract void updateFile(@NotNull Path file);

    /**
     * Recreates the given file, first deleting it if present and then creating it anew.
     *
     * @param file the file to update
     */
    public abstract void updateFile(@NotNull File file);

    /**
     * Renames the given file.
     *
     * @param file    the file to rename
     * @param newName the new name of the file
     */
    public abstract void rename(@NotNull File file, @NotNull String newName);

    /**
     * Creates the given directory, including any missing parent directories.
     *
     * @param path the directory to create, or {@code null} to do nothing
     */
    public abstract void createDirectory(@Nullable Path path);

    /**
     * Copies a single file from one location to another.
     *
     * @param from   the path of the file to copy
     * @param target the destination path
     */
    public abstract void doCopy(@NotNull String from, @NotNull String target);

    /**
     * Deletes every file directly contained in the given directory, without recursing into
     * subdirectories.
     *
     * @param dirPath the directory whose files should be deleted
     */
    public abstract void deleteAllFilesInDirectory(@NotNull Path dirPath);

    /**
     * Recursively deletes the given directory and all of its contents.
     *
     * @param dirPath the directory to delete
     */
    public abstract void deleteDirectory(@NotNull Path dirPath);

    /**
     * Deletes the given directory (or file) if it exists, and then recreates it as an empty
     * directory.
     *
     * @param path the directory to recreate
     */
    public abstract void recreateDirectory(@NotNull Path path);

    /**
     * Recursively copies a directory's contents to another location.
     *
     * @param path   the source directory
     * @param target the destination directory
     */
    public abstract void copyDirectory(@NotNull Path path, @NotNull Path target);

    /**
     * Recursively copies a directory's contents to another location, skipping any file whose name
     * is contained in {@code excludedFiles}.
     *
     * @param path          the source directory
     * @param target        the destination directory
     * @param excludedFiles the file names to skip while copying
     */
    public abstract void copyDirectory(@NotNull Path path, @NotNull Path target, @NotNull Collection<String> excludedFiles);

}

File: python/database-driver-api/src/database_driver/api/json/file/file_provider.py
"""Mirror of ``de.lino.database.json.file.FileProvider``."""

from __future__ import annotations

import os
from abc import ABC, abstractmethod
from collections.abc import Collection
from typing import ClassVar

PathLike = str | os.PathLike[str]


class FileProvider(ABC):
    """Abstraction over the filesystem operations used throughout the driver (creating,
    deleting, renaming and copying files and directories), so that ``JsonDocument`` and
    the various providers do not depend on a concrete filesystem implementation directly.

    A single, globally accessible instance is installed via :meth:`set_instance` and
    retrieved through :meth:`get_instance`.

    The Java edition carries ``File``/``Path`` overload pairs for several operations;
    Python needs only one method per operation, since every path parameter accepts any
    ``os.PathLike`` alongside ``str``.
    """

    _instance: ClassVar[FileProvider | None] = None

    @classmethod
    def get_instance(cls) -> FileProvider | None:
        """Returns the globally accessible instance, or ``None`` if none was installed."""
        return FileProvider._instance

    @classmethod
    def set_instance(cls, instance: FileProvider) -> None:
        """Installs the given provider as the globally accessible instance returned by
        :meth:`get_instance`."""
        FileProvider._instance = instance

    @abstractmethod
    def delete_file(self, file: PathLike) -> None:
        """Deletes the given file, if it exists."""

    @abstractmethod
    def create_file(self, file: PathLike) -> None:
        """Creates the given file if it does not exist yet."""

    @abstractmethod
    def update_file(self, file: PathLike) -> None:
        """Recreates the given file: deletes it if present, then creates it anew."""

    @abstractmethod
    def rename(self, file: PathLike, new_name: str) -> None:
        """Renames the given file to ``new_name`` within its directory."""

    @abstractmethod
    def create_directory(self, path: PathLike | None) -> None:
        """Creates the given directory (and any missing parents), if it does not exist."""

    @abstractmethod
    def do_copy(self, from_path: str, target: str) -> None:
        """Copies the file at ``from_path`` to ``target``, replacing any existing file."""

    @abstractmethod
    def delete_all_files_in_directory(self, dir_path: PathLike) -> None:
        """Deletes every file directly inside the given directory, leaving the directory
        itself in place."""

    @abstractmethod
    def delete_directory(self, dir_path: PathLike) -> None:
        """Deletes the given directory and everything below it."""

    @abstractmethod
    def recreate_directory(self, path: PathLike) -> None:
        """Deletes the given directory if present, then creates it empty."""

    @abstractmethod
    def copy_directory(self, path: PathLike, target: PathLike, excluded_files: Collection[str] = ()) -> None:
        """Recursively copies the directory at ``path`` to ``target``, skipping any file
        whose name is listed in ``excluded_files``."""
